parser/src/soft_keywords.rs

```diff
@@ -43,44 +46,81 @@ where
+                            match tok {
+                                Tok::Newline => break,
+                                Tok::Lambda if nesting == 0 => seen_lambda = true,
+                                Tok::Colon if nesting == 0 => {
+                                    if seen_lambda {
+                                        seen_lambda = false;
+                                    } else if !first {
+                                        seen_colon = true;
+                                    }
                                 }
+                                Tok::Lpar | Tok::Lsqb | Tok::Lbrace => nesting += 1,
+                                Tok::Rpar | Tok::Rsqb | Tok::Rbrace => nesting -= 1,
+                                _ => {}
                             }
-                            Tok::Lpar | Tok::Lsqb | Tok::Lbrace => nesting += 1,
-                            Tok::Rpar | Tok::Rsqb | Tok::Rbrace => nesting -= 1,
-                            _ => {}
+                            first = false;
+                        }
+                        if !seen_colon {
+                            next = Some(Ok((soft_to_name(tok), *range)));
                         }
-                        first = false;
                     }
-                    if !seen_colon {
+                }
+                // For `type` all of the following conditions must be met:
+                // 1. The token is at the start of a logical line.
+                // 2. The type token is immediately followed by a name token.
+                // 3. The name token is eventually followed by an equality token.
+                Tok::Type => {
+                    if !self.start_of_line {
                         next = Some(Ok((soft_to_name(tok), *range)));
+                    } else {
+                        let mut nesting = 0;
+                        let mut first = true;
+                        let mut seen_name = false;
+                        let mut seen_equal = false;
+                        while let Some(Ok((tok, _))) = self.underlying.peek() {
+                            match tok {
+                                Tok::Newline => break,
+                                Tok::Name { .. } |
+                                    // We treat a soft keyword token following a type token as a
+                                    // name to support cases like `type type = int` or `type match = int`
+                                    Tok::Type | Tok::Match | Tok::Case
+                                    if first => seen_name = true,
+                                Tok::Equal if nesting == 0 && seen_name => seen_equal = true,
+                                Tok::Lpar | Tok::Lsqb | Tok::Lbrace => nesting += 1,
+                                Tok::Rpar | Tok::Rsqb | Tok::Rbrace => nesting -= 1,
+                                _ => {}
+                            }
+                            first = false;
```

Comment:
Since we _need_ the first token to have certain conditions, I might suggest checking that first, and skipping this loop entirely if it's not a name (or, alternatively, finding a way to break after checking the first token, if it's a non-name). Otherwise, we're iterating over the whole line unnecessarily.

Something like:

```rust
let mut is_type_alias = false;
if let Some(Ok((tok, _))) = self.underlying.peek() {
    // The next token is a name, type, match, or case.
    if matches!(tok, Tok::Name { .. } | Tok::Type | Tok::Match | Tok::Case) {
        // The name is followed, at some point, by an equal sign.
        let mut nesting = 0;
        while let Some(Ok((tok, _))) = self.underlying.peek() {
            match tok {
                Tok::Newline => break,
                Tok::Equal if nesting == 0 && seen_name => {
                    is_type_alias = true;
                    break;
                }
                Tok::Lpar | Tok::Lsqb | Tok::Lbrace => nesting += 1,
                Tok::Rpar | Tok::Rsqb | Tok::Rbrace => nesting -= 1,
                _ => {}
            }
        }
    }
}

if !is_type_alias {
    next = Some(Ok((soft_to_name(tok), *range)));
}
```

Reply:
Ah yeah makes sense to avoid iterating over the line!

Reply:
https://github.com/RustPython/Parser/pull/97/commits/e7ac2eff1e5e6fe4b6cd0019ba9e816ca03e304a